FLT/CrazyFrobenius.lean

```diff
@@ -0,0 +1,986 @@
+    exact congrArg (map (algebraMap B (B ⧸ Q))) h
+  apply m_mapsto_F
+
+lemma m_expand_char_q : (Polynomial.map (algebraMap A (A ⧸ P)) m) ^ q =
+    (expand _ q (Polynomial.map (algebraMap A (A ⧸ P)) m)) := by
+  apply pow_eq_expand
+
+lemma B_m_expand_char_q : (Polynomial.map (algebraMap A (B ⧸ Q)) m) ^ q =
+    (expand _ q (Polynomial.map (algebraMap A (B ⧸ Q)) m)) := by
+  have st : (algebraMap A (B ⧸ Q)) =
+      RingHom.comp (algebraMap (A ⧸ P) (B ⧸ Q)) (algebraMap A (A ⧸ P)) := by
+    exact IsScalarTower.algebraMap_eq A (A ⧸ P)  (B ⧸ Q)
+  rw[st]
+  rw [← Polynomial.map_map, ← Polynomial.map_pow]
+  rw[m_expand_char_q, map_expand]
+
+lemma A_B_scalar_tower_pow_m : (Polynomial.map (algebraMap A (B ⧸ Q)) m) ^ q =
+    (( Polynomial.map ( RingHom.comp (algebraMap B (B ⧸ Q)) (algebraMap A B))) m) ^ q := by
+  have st : (algebraMap A (B ⧸ Q)) =
+      RingHom.comp (algebraMap B (B ⧸ Q)) (algebraMap A B) := by
+    exact IsScalarTower.algebraMap_eq A B (B ⧸ Q)
+  rw[st]
+
+lemma A_B_scalar_tower_m :  (Polynomial.map (algebraMap A (B ⧸ Q)) m) =
+    (( Polynomial.map ( RingHom.comp (algebraMap B (B ⧸ Q)) (algebraMap A B))) m) := by
+  have st : (algebraMap A (B ⧸ Q)) =
+      RingHom.comp (algebraMap B (B ⧸ Q)) (algebraMap A B) := by
+    exact IsScalarTower.algebraMap_eq A B (B ⧸ Q)
+  rw[st]
+
+lemma pow_expand_A_B_scalar_tower_m :
+    (( Polynomial.map ( RingHom.comp (algebraMap B (B ⧸ Q)) (algebraMap A B))) m) ^ q =
+    (expand _ q ( Polynomial.map ( RingHom.comp (algebraMap B (B ⧸ Q)) (algebraMap A B) ) m)) := by
+  have h : (Polynomial.map (algebraMap A (B ⧸ Q)) m) ^ q =
+    (expand _ q (Polynomial.map (algebraMap A (B ⧸ Q)) m)) := B_m_expand_char_q P Q_ne_bot
+  rw[← A_B_scalar_tower_m]
+  exact h
+
+lemma pow_expand_A_B_scalar_tower_F :
+    ( Polynomial.map (algebraMap B (B ⧸ Q)) F) ^ q =
+    (expand _ q ( Polynomial.map (algebraMap B (B ⧸ Q)) F)) := by
+  have h :  (( Polynomial.map ( RingHom.comp (algebraMap B (B ⧸ Q)) (algebraMap A B))) m) ^ q =
+    (expand _ q ( Polynomial.map ( RingHom.comp (algebraMap B (B ⧸ Q)) (algebraMap A B) ) m)) := by
+      exact pow_expand_A_B_scalar_tower_m P Q_ne_bot
+  rw[← m_mapsto_F]
+  rw[Polynomial.map_map]
+  exact h
+
+lemma F_expand_eval_eq_eval_pow :
+    (eval₂ (Ideal.Quotient.mk Q) (Ideal.Quotient.mk Q β) F) ^ q =
+    (eval₂ (Ideal.Quotient.mk Q) (Ideal.Quotient.mk Q (β ^ q)) F) := by
+  simp only [← Polynomial.eval_map,  ←  Ideal.Quotient.algebraMap_eq, ← Polynomial.coe_evalRingHom,
+    ← map_pow, pow_expand_A_B_scalar_tower_F]
+  simp only [Ideal.Quotient.algebraMap_eq, coe_evalRingHom, expand_eval, map_pow]
+
+lemma quotient_F_is_product_of_quot :
+    (Polynomial.map (Ideal.Quotient.mk Q) F) =
+    ∏ τ : L ≃ₐ[K] L, (X - C ((Ideal.Quotient.mk Q) ((galRestrict A K L B τ) β))) := by
+  rw[← Polynomial.coe_mapRingHom]
+  erw[map_prod]
+  simp only [map_sub, coe_mapRingHom, map_X, map_C]
+
+lemma quotient_F_is_root_iff_is_conjugate (x : (B ⧸ Q)) :
```

Comment:
It seems to me that many lemmas and definitions until here would apply to any element in `B` instead of the carefully chosen `β`. It may be worth generalizing the construction of `F`, etc, in that generality.